fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+from fl4health.losses.mkmmd_loss import MkMmdLoss
+from fl4health.model_bases.moon_base import MoonModel
+from fl4health.parameter_exchange.full_exchanger import FullParameterExchanger
+from fl4health.utils.losses import EvaluationLosses, LossMeterType, TrainingLosses
+from fl4health.utils.metrics import Metric
+
+
+class DittoClient(BasicClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        lam: float = 1.0,
+        mkmmd_loss_weight: float = 10.0,
+        beta_update_interval: int = 20,
+        feature_l2_norm: Optional[float] = 0.0,
+    ) -> None:
+        """
+        This client implements the Ditto algorithm from Ditto: Fair and Robust Federated Learning Through
+        Personalization. The idea is that we want to train personalized versions of the global model for each client.
+        So we simultaneously train a global model that is aggregated on the server-side and use those weights to also
+        constrain the training of a local model. The constraint for this local model is identical to the FedProx loss.
+
+        Args:
+            data_path (Path): path to the data to be used to load the data for client-side training
+            metrics (Sequence[Metric]): Metrics to be computed based on the labels and predictions of the client model
+            device (torch.device): Device indicator for where to send the model, batches, labels etc. Often 'cpu' or
+                'cuda'
+            loss_meter_type (LossMeterType, optional): Type of meter used to track and compute the losses over
+                each batch. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
+                checkpointing. Defaults to None.
+            metrics_reporter (Optional[MetricsReporter], optional): A metrics reporter instance to record the metrics
+                during the execution. Defaults to an instance of MetricsReporter with default init parameters.
+            lam (float, optional): weight applied to the Ditto drift loss. Defaults to 1.0.
+            mkmmd_loss_weight (float, optional): weight applied to the MK-MMD loss. Defaults to 10.0.
+            beta_update_interval (int, optional): interval at which to update the betas for the MK-MMD loss.
+                Defaults to 20.
+            feature_l2_norm (Optional[float], optional): weight applied to the L2 norm of the features.
+            Defaults to 0.0.
+        """
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.lam = lam
+        self.mkmmd_loss_weight = mkmmd_loss_weight
```

Comment:
Perhaps we can include a small warning here if the mkmmd weight is 0 that the loss function won't be used and the individual should consider using the vanilla `DittoClient` instead?